### PR TITLE
Save picture in S3 and return S3 URL

### PR DESCRIPTION
**Pull request by alexdesi**
Jira: https://dsdmoj.atlassian.net/browse/P4-1143

## Proposed Changes

When the Image is retrieved from Nomis, it store it on S3 and return the URL

## To test/demo change

call the API get `/people/person_id/image`

## Things to check & link
- [ ] API docs has been updated if necessary
- [ ] New environment variables have been added to staging configuration

### CODE REVIEW

File: app/controllers/api/v1/people_controller.rb [hunk 1]
```diff
@@ -25,13 +25,13 @@ def update
...
-        if image_data
-          send_data image_data, type: 'image/jpg', disposition: 'inline'
-        else
-          render status: :not_found
+        if !person.picture.attached?
+          image_blob = NomisClient::Image::get(person)
+          person.attach_picture(image_blob)
         end
+
+        render json: { url: url_for(person.picture) }
```
Comment (starswan) [inline app/controllers/api/v1/people_controller.rb]: don't think this is right. I think JSON API spec requires a data: element with an array in it

File: app/lib/nomis_client/image.rb [hunk 2]
```diff
@@ -5,7 +5,9 @@ module NomisClient
   # :nocov:
   class Image < NomisClient::Base
     class << self
-      def get booking_id
+      def get(person)
+        booking_id = person.latest_profile.latest_nomis_booking_id
```
Comment (starswan) [inline app/lib/nomis_client/image.rb]: not happy with this change - it ties the Nomis::Image client to our person model (additional coupling) - all our other Nomis::<x> modules are independent, so this is inconsistent as well

File: app/models/person.rb [hunk 3]
```diff
@@ -4,7 +4,22 @@ class Person < VersionedModel
   has_many :profiles, dependent: :destroy
   has_many :moves, dependent: :destroy
 
+  has_one_attached :picture
+
   def latest_profile
     profiles.last
   end
+
+  def attach_picture(image_blob)
```
Comment (starswan) [inline app/models/person.rb]: was it not possible to use a StringIO object here? If not, I think an intention comment would be useful

File: spec/requests/api/v1/people_controller_image_spec.rb [hunk 4]
```diff
@@ -13,28 +13,46 @@
         allow(NomisClient::Image).to receive(:get).and_return(nil)
       end
 
-      it 'returns 404' do
-        get "/api/v1/people/#{person.id}/image", params: { access_token: token.token }
+      it 'returns resource not found 404' do
```
Comment (starswan) [inline spec/requests/api/v1/people_controller_image_spec.rb]: This was a different test - to check that if the person didn't have an image then we return 404. The non-existant-id case should IMHO be a new test

File: app/models/image.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class Image
+  include ActiveModel::Serialization
+
+  attr_accessor :id, :url
```
Comment (starswan) [inline app/models/image.rb]: these should be attr_readers

File: spec/requests/api/v1/people_controller_image_spec.rb [hunk 6]
```diff
@@ -13,28 +13,47 @@
...
         expect(response).to be_successful
       end
 
-      it 'returns the image' do
-        expect(response.body).to eq(image_data)
+      it 'contains the url of the image' do
+        get_image
+
+
+        expect(JSON.parse(response.body)['data']['attributes']['url']).to include person.id + '.jpg'
```
Comment (starswan) [inline spec/requests/api/v1/people_controller_image_spec.rb]: IMHO we can just check for equality of the JSON reply - it's not too big or onerous

File: app/models/image.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class Image
+  include ActiveModel::Serialization
+
```
Comment (starswan) [inline app/models/image.rb]: I'm a little unsure about this behaviour. You're claiming that this is an 'image' object, but its ID field is that of a 'person' object. Doesn't feel quite right?

File: app/controllers/api/v1/people_controller.rb (hunk 1 above)
Comment (alexdesi) [reply app/controllers/api/v1/people_controller.rb]: Fixed, this is now using a serializer.

File: app/models/person.rb (hunk 3 above)
Comment (alexdesi) [reply app/models/person.rb]: I think I got an error about the format, I think it's clear from the fact we handle an image here that it is a blob, Binary Large Object.

File: app/models/image.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class Image
+  include ActiveModel::Serialization
```
Comment (starswan) [inline app/models/image.rb]: As we have an ImageSerializer, I don't think this line is required. This class is just a property holder

File: app/serializers/person_serializer.rb [hunk 9]
```diff
@@ -13,8 +13,6 @@ class PersonSerializer < ActiveModel::Serializer
     :gender_additional_information,
   )
```
Comment (starswan) [inline app/serializers/person_serializer.rb]: This deletion introduces new behaviour. It's now not possible to tell if a person should have an image property or not. I think we should therefore have a test to show that if we don't have a prison_number on the person, then /people/<id>/images behaves correctly (which I think means 404)

File: app/lib/nomis_client/image.rb (hunk 2 above)
Comment (alexdesi) [reply app/lib/nomis_client/image.rb]: I got your point, it was an attempt to clean a bit the controller,
since that's a query, I'll move it in the model.

Comment (starswan) [reply app/lib/nomis_client/image.rb]: This isnt a model - it's a service to get images from NOMIS

File: spec/requests/api/v1/people_controller_image_spec.rb (hunk 4 above)
Comment (alexdesi) [reply spec/requests/api/v1/people_controller_image_spec.rb]: fixed, I've created a different context for that.

File: spec/requests/api/v1/people_controller_image_spec.rb (hunk 6 above)
Comment (alexdesi) [reply spec/requests/api/v1/people_controller_image_spec.rb]: that would imply to generate in full the URL, I think this test works and it is explicit

File: spec/requests/api/v1/people_controller_image_spec.rb (hunk 4 above)
Comment (starswan) [reply spec/requests/api/v1/people_controller_image_spec.rb]: we still seem to have 2 deletions according to github?

File: app/models/image.rb (hunk 8 above)
Comment (alexdesi) [reply app/models/image.rb]: It is required, if you do not put it you get

```
 NoMethodError: undefined method `read_attribute_for_serialization' 
```

File: app/models/image.rb (hunk 7 above)
Comment (alexdesi) [reply app/models/image.rb]: I see you point, but the initialize helps to understand what actually the id is:

```
def initialize(person_id, image_url)
    @id = person_id
    @url = image_url
  end
```

Image is an entity per se, so it's fine to have it as a model imo,
the models folders do not necessary need to contain only record model, there are other classes not activerecord.

File: app/serializers/person_serializer.rb (hunk 9 above)
Comment (alexdesi) [reply app/serializers/person_serializer.rb]: This was discussed with @teneightfive, it complicates the FE logic, we preferred to simplify the logic removing it,
We might want to put it back in the future, so FE would need to retrieve the images only for the people that do not have yet URL,

File: app/lib/nomis_client/image.rb (hunk 2 above)
Comment (alexdesi) [reply app/lib/nomis_client/image.rb]: I've created a services for that.

**Pull request merged**